File: server/srs/fsrs.py
"""A self-contained FSRS-6 scheduler.

FSRS (Free Spaced Repetition Scheduler) is the modern, benchmarked SRS the
DEEP_SEARCH blueprint selects over SM-2/Leitner. This is a faithful, dependency-
free implementation of FSRS-6 (21 parameters, w0..w20) with the standard learning
/ relearning step machine - the same behaviour as py-fsrs, kept in-repo so the
scheduling logic is inspectable and version-stable.

The scheduler is pure: it takes a card's memory state + a rating + a timestamp and
returns the next state. Persistence (Card / ReviewLog rows) lives in models.py; the
per-user `desired_retention` comes from the profile.

Rating scale (Anki/FSRS standard): 1=Again, 2=Hard, 3=Good, 4=Easy.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime, timedelta

# States (match Anki/py-fsrs).
NEW = 0
LEARNING = 1
REVIEW = 2
RELEARNING = 3

AGAIN, HARD, GOOD, EASY = 1, 2, 3, 4

# FSRS-6 default parameters (open-spaced-repetition, 21 weights w0..w20).
DEFAULT_PARAMETERS: tuple[float, ...] = (
    0.2172,
    1.1771,
    3.2602,
    16.1507,
    7.0114,
    0.57,
    2.0966,
    0.0069,
    1.5261,
    0.112,
    1.0178,
    1.849,
    0.1133,
    0.3127,
    2.2934,
    0.2191,
    3.0004,
    0.7536,
    0.3332,
    0.1437,
    0.2,
)

# Sub-day steps before a card graduates to day-granularity review intervals.
LEARNING_STEPS = (timedelta(minutes=1), timedelta(minutes=10))
RELEARNING_STEPS = (timedelta(minutes=10),)
MAX_INTERVAL_DAYS = 36500
MIN_STABILITY = 0.01
STABILITY_MIN_CLAMP = 0.01


@dataclass
class MemoryState:
    """The card fields the scheduler reads and writes. Mirrors the Card model."""

    state: int = NEW
    step: int | None = None  # index into LEARNING/RELEARNING steps, None once in review
    stability: float | None = None
    difficulty: float | None = None
    due: datetime | None = None
    last_review: datetime | None = None


def _clamp(x: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, x))


class FSRS:
    def __init__(
        self,
        parameters: tuple[float, ...] | list[float] | None = None,
        desired_retention: float = 0.9,
    ):
        self.w = tuple(parameters or DEFAULT_PARAMETERS)
        if len(self.w) != 21:
            raise ValueError("FSRS-6 requires exactly 21 parameters")
        self.desired_retention = desired_retention
        self.decay = -self.w[20]
        self.factor = 0.9 ** (1.0 / self.decay) - 1.0

    # ── forgetting curve ───────────────────────────────────────────────────────

    def retrievability(self, elapsed_days: float, stability: float) -> float:
        elapsed_days = max(0.0, elapsed_days)
        return (1.0 + self.factor * elapsed_days / stability) ** self.decay

    def interval_days(self, stability: float) -> int:
        raw = (stability / self.factor) * (self.desired_retention ** (1.0 / self.decay) - 1.0)
        return int(_clamp(round(raw), 1, MAX_INTERVAL_DAYS))

    # ── initial state ────────────────────────────────────────────────────────

    def _initial_stability(self, rating: int) -> float:
        return max(self.w[rating - 1], MIN_STABILITY)

    def _initial_difficulty(self, rating: int) -> float:
        return _clamp(self.w[4] - math.exp(self.w[5] * (rating - 1)) + 1.0, 1.0, 10.0)

    # ── updates ────────────────────────────────────────────────────────────────

    def _next_difficulty(self, d: float, rating: int) -> float:
        delta = -self.w[6] * (rating - 3)
        d_prime = d + delta * (10.0 - d) / 9.0  # linear damping
        # mean-reversion toward the difficulty of an initial "Easy" (rating 4)
        reverted = self.w[7] * self._initial_difficulty(EASY) + (1.0 - self.w[7]) * d_prime
        return _clamp(reverted, 1.0, 10.0)

    def _short_term_stability(self, stability: float, rating: int) -> float:
        sinc = math.exp(self.w[17] * (rating - 3 + self.w[18])) * stability ** (-self.w[19])
        if rating >= GOOD:
            sinc = max(sinc, 1.0)
        return _clamp(stability * sinc, STABILITY_MIN_CLAMP, MAX_INTERVAL_DAYS)

    def _next_recall_stability(self, d: float, s: float, r: float, rating: int) -> float:
        hard_penalty = self.w[15] if rating == HARD else 1.0
        easy_bonus = self.w[16] if rating == EASY else 1.0
        new_s = s * (
            1.0
            + math.exp(self.w[8])
            * (11.0 - d)
            * (s ** -self.w[9])
            * (math.exp(self.w[10] * (1.0 - r)) - 1.0)
            * hard_penalty
            * easy_bonus
        )
        return _clamp(new_s, STABILITY_MIN_CLAMP, MAX_INTERVAL_DAYS)

    def _next_forget_stability(self, d: float, s: float, r: float) -> float:
        new_s = (
            self.w[11]
            * (d ** -self.w[12])
            * (((s + 1.0) ** self.w[13]) - 1.0)
            * math.exp(self.w[14] * (1.0 - r))
        )
        # A lapse must not raise stability above where it was.
        return _clamp(min(new_s, s), STABILITY_MIN_CLAMP, MAX_INTERVAL_DAYS)

    # ── the review transition ───────────────────────────────────────────────────

    def review(self, card: MemoryState, rating: int, now: datetime) -> MemoryState:
        """Return the card's next memory state after `rating` at time `now`."""
        elapsed_days = 0.0
        if card.last_review is not None:
            elapsed_days = max(0.0, (now - card.last_review).total_seconds() / 86400.0)

        first_review = card.stability is None or card.difficulty is None
        if first_review:
            stability = self._initial_stability(rating)
            difficulty = self._initial_difficulty(rating)
        else:
            r = self.retrievability(elapsed_days, card.stability)
            difficulty = self._next_difficulty(card.difficulty, rating)
            if elapsed_days < 1.0:
                stability = self._short_term_stability(card.stability, rating)
            elif rating == AGAIN:
                stability = self._next_forget_stability(difficulty, card.stability, r)
            else:
                stability = self._next_recall_stability(difficulty, card.stability, r, rating)

        state, step, interval = self._schedule(card, rating, stability)

        return MemoryState(
            state=state,
            step=step,
            stability=stability,
            difficulty=difficulty,
            due=now + interval,
            last_review=now,
        )

    def _schedule(self, card: MemoryState, rating: int, stability: float):
        """Decide the next (state, step, due-interval) using the step machine."""
        cur_state = card.state if card.state != NEW else LEARNING
        step = card.step or 0

        if cur_state in (LEARNING, NEW):
            return self._step_through(LEARNING_STEPS, step, rating, stability, RELEARNING_STEPS)
        if cur_state == RELEARNING:
            return self._step_through(RELEARNING_STEPS, step, rating, stability, RELEARNING_STEPS)

        # REVIEW state
        if rating == AGAIN and RELEARNING_STEPS:
            return RELEARNING, 0, RELEARNING_STEPS[0]
        return REVIEW, None, timedelta(days=self.interval_days(stability))

    def _step_through(self, steps, step, rating, stability, relearn_steps):
        """Learning/relearning step progression (mirrors py-fsrs)."""
        if not steps:
            return REVIEW, None, timedelta(days=self.interval_days(stability))

        if rating == AGAIN:
            return (LEARNING if steps is LEARNING_STEPS else RELEARNING), 0, steps[0]
        if rating == HARD:
            state = LEARNING if steps is LEARNING_STEPS else RELEARNING
            if step == 0 and len(steps) == 1:
                return state, 0, steps[0] * 1.5
            if step == 0 and len(steps) >= 2:
                return state, 0, (steps[0] + steps[1]) / 2
            return state, step, steps[min(step, len(steps) - 1)]
        if rating == GOOD:
            if step + 1 >= len(steps):
                return REVIEW, None, timedelta(days=self.interval_days(stability))
            state = LEARNING if steps is LEARNING_STEPS else RELEARNING
            return state, step + 1, steps[step + 1]
        # EASY graduates immediately
        return REVIEW, None, timedelta(days=self.interval_days(stability))

File: app/lib/srs/fsrs.dart
/// A faithful Dart port of the server's FSRS-6 scheduler (server/srs/fsrs.py).
///
/// Offline devices schedule reviews locally with this port while the server
/// stays the canonical store; the two implementations MUST stay bit-compatible
/// or schedules would drift between devices. Parity is locked by shared test
/// vectors (scripts/gen_fsrs_vectors.py → test/srs/fsrs_parity_test.dart);
/// [fsrsVectorsVersion] must match the fixture's version - if the server
/// scheduler changes, regenerate the vectors AND update this port in the same
/// change.
///
/// Rating scale (Anki/FSRS standard): 1=Again, 2=Hard, 3=Good, 4=Easy.
library;

import 'dart:math' as math;

/// Version of the shared parity vectors this port was verified against.
const int fsrsVectorsVersion = 1;

// States (match Anki/py-fsrs and srs/models.py).
const int stateNew = 0;
const int stateLearning = 1;
const int stateReview = 2;
const int stateRelearning = 3;

const int ratingAgain = 1;
const int ratingHard = 2;
const int ratingGood = 3;
const int ratingEasy = 4;

/// FSRS-6 default parameters (open-spaced-repetition, 21 weights w0..w20).
const List<double> defaultParameters = [
  0.2172,
  1.1771,
  3.2602,
  16.1507,
  7.0114,
  0.57,
  2.0966,
  0.0069,
  1.5261,
  0.112,
  1.0178,
  1.849,
  0.1133,
  0.3127,
  2.2934,
  0.2191,
  3.0004,
  0.7536,
  0.3332,
  0.1437,
  0.2,
];

/// Sub-day steps before a card graduates to day-granularity review intervals.
const List<Duration> learningSteps = [Duration(minutes: 1), Duration(minutes: 10)];
const List<Duration> relearningSteps = [Duration(minutes: 10)];
const int maxIntervalDays = 36500;
const double minStability = 0.01;
const double stabilityMinClamp = 0.01;

/// The card fields the scheduler reads and writes. Mirrors the server Card row.
class MemoryState {
  const MemoryState({
    this.state = stateNew,
    this.step,
    this.stability,
    this.difficulty,
    this.due,
    this.lastReview,
  });

  final int state;

  /// Index into learning/relearning steps, null once in review.
  final int? step;
  final double? stability;
  final double? difficulty;
  final DateTime? due;
  final DateTime? lastReview;
}

double _clamp(double x, double lo, double hi) => math.max(lo, math.min(hi, x));

/// Python's round() rounds half to even (banker's rounding); Dart's .round()
/// rounds half away from zero. interval_days must round exactly like the
/// server. Only called with positive values.
int _pyRound(double x) {
  final floor = x.floorToDouble();
  final diff = x - floor;
  if (diff > 0.5) return floor.toInt() + 1;
  if (diff < 0.5) return floor.toInt();
  final f = floor.toInt();
  return f.isEven ? f : f + 1;
}

class Fsrs {
  Fsrs({List<double>? parameters, this.desiredRetention = 0.9})
      : w = List.unmodifiable(parameters ?? defaultParameters) {
    if (w.length != 21) {
      throw ArgumentError('FSRS-6 requires exactly 21 parameters');
    }
    decay = -w[20];
    factor = math.pow(0.9, 1.0 / decay) - 1.0;
  }

  final List<double> w;
  final double desiredRetention;
  late final double decay;
  late final double factor;

  // ── forgetting curve ───────────────────────────────────────────────────────

  double retrievability(double elapsedDays, double stability) {
    final elapsed = math.max(0.0, elapsedDays);
    return math.pow(1.0 + factor * elapsed / stability, decay).toDouble();
  }

  int intervalDays(double stability) {
    final raw = (stability / factor) *
        (math.pow(desiredRetention, 1.0 / decay).toDouble() - 1.0);
    final rounded = _pyRound(raw);
    if (rounded < 1) return 1;
    if (rounded > maxIntervalDays) return maxIntervalDays;
    return rounded;
  }

  // ── initial state ────────────────────────────────────────────────────────

  double _initialStability(int rating) => math.max(w[rating - 1], minStability);

  double _initialDifficulty(int rating) =>
      _clamp(w[4] - math.exp(w[5] * (rating - 1)) + 1.0, 1.0, 10.0);

  // ── updates ────────────────────────────────────────────────────────────────

  double _nextDifficulty(double d, int rating) {
    final delta = -w[6] * (rating - 3);
    final dPrime = d + delta * (10.0 - d) / 9.0; // linear damping
    // mean-reversion toward the difficulty of an initial "Easy" (rating 4)
    final reverted = w[7] * _initialDifficulty(ratingEasy) + (1.0 - w[7]) * dPrime;
    return _clamp(reverted, 1.0, 10.0);
  }

  double _shortTermStability(double stability, int rating) {
    var sinc = math.exp(w[17] * (rating - 3 + w[18])) *
        math.pow(stability, -w[19]).toDouble();
    if (rating >= ratingGood) sinc = math.max(sinc, 1.0);
    return _clamp(stability * sinc, stabilityMinClamp, maxIntervalDays.toDouble());
  }

  double _nextRecallStability(double d, double s, double r, int rating) {
    final hardPenalty = rating == ratingHard ? w[15] : 1.0;
    final easyBonus = rating == ratingEasy ? w[16] : 1.0;
    final newS = s *
        (1.0 +
            math.exp(w[8]) *
                (11.0 - d) *
                math.pow(s, -w[9]).toDouble() *
                (math.exp(w[10] * (1.0 - r)) - 1.0) *
                hardPenalty *
                easyBonus);
    return _clamp(newS, stabilityMinClamp, maxIntervalDays.toDouble());
  }

  double _nextForgetStability(double d, double s, double r) {
    final newS = w[11] *
        math.pow(d, -w[12]).toDouble() *
        (math.pow(s + 1.0, w[13]).toDouble() - 1.0) *
        math.exp(w[14] * (1.0 - r));
    // A lapse must not raise stability above where it was.
    return _clamp(math.min(newS, s), stabilityMinClamp, maxIntervalDays.toDouble());
  }

  // ── the review transition ───────────────────────────────────────────────────

  /// Return the card's next memory state after [rating] at time [now].
  MemoryState review(MemoryState card, int rating, DateTime now) {
    var elapsedDays = 0.0;
    if (card.lastReview != null) {
      // Mirrors Python's timedelta.total_seconds()/86400 (µs → s, then days).
      elapsedDays = math.max(
          0.0, now.difference(card.lastReview!).inMicroseconds / 1e6 / 86400.0);
    }

    final firstReview = card.stability == null || card.difficulty == null;
    double stability;
    double difficulty;
    if (firstReview) {
      stability = _initialStability(rating);
      difficulty = _initialDifficulty(rating);
    } else {
      final r = retrievability(elapsedDays, card.stability!);
      difficulty = _nextDifficulty(card.difficulty!, rating);
      if (elapsedDays < 1.0) {
        stability = _shortTermStability(card.stability!, rating);
      } else if (rating == ratingAgain) {
        stability = _nextForgetStability(difficulty, card.stability!, r);
      } else {
        stability = _nextRecallStability(difficulty, card.stability!, r, rating);
      }
    }

    final (state, step, interval) = _schedule(card, rating, stability);

    return MemoryState(
      state: state,
      step: step,
      stability: stability,
      difficulty: difficulty,
      due: now.add(interval),
      lastReview: now,
    );
  }

  /// Decide the next (state, step, due-interval) using the step machine.
  (int, int?, Duration) _schedule(MemoryState card, int rating, double stability) {
    final curState = card.state != stateNew ? card.state : stateLearning;
    final step = card.step ?? 0;

    if (curState == stateLearning || curState == stateNew) {
      return _stepThrough(learningSteps, step, rating, stability, learning: true);
    }
    if (curState == stateRelearning) {
      return _stepThrough(relearningSteps, step, rating, stability, learning: false);
    }

    // REVIEW state
    if (rating == ratingAgain && relearningSteps.isNotEmpty) {
      return (stateRelearning, 0, relearningSteps[0]);
    }
    return (stateReview, null, Duration(days: intervalDays(stability)));
  }

  /// Learning/relearning step progression (mirrors py-fsrs).
  (int, int?, Duration) _stepThrough(
    List<Duration> steps,
    int step,
    int rating,
    double stability, {
    required bool learning,
  }) {
    if (steps.isEmpty) {
      return (stateReview, null, Duration(days: intervalDays(stability)));
    }

    final state = learning ? stateLearning : stateRelearning;
    if (rating == ratingAgain) {
      return (state, 0, steps[0]);
    }
    if (rating == ratingHard) {
      if (step == 0 && steps.length == 1) return (state, 0, steps[0] * 1.5);
      if (step == 0 && steps.length >= 2) return (state, 0, (steps[0] + steps[1]) ~/ 2);
      return (state, step, steps[math.min(step, steps.length - 1)]);
    }
    if (rating == ratingGood) {
      if (step + 1 >= steps.length) {
        return (stateReview, null, Duration(days: intervalDays(stability)));
      }
      return (state, step + 1, steps[step + 1]);
    }
    // EASY graduates immediately
    return (stateReview, null, Duration(days: intervalDays(stability)));
  }
}
